src/eventer/eventer_epoll_impl.c

```diff
@@ -285,9 +296,24 @@ static void eventer_epoll_impl_trigger(eventer_t e, int mask) {
     e->mask = 0;
     needs_add = 1;
   }
-  if(e != master_fds[fd].e) return;
+  if(e != master_fds[fd].e) {
+    mtevL(eventer_deb, "Incoming event: %p, does not match master list: %p\n", e, master_fds[fd].e);
+    return;
+  }
   lockstate = acquire_master_fd(fd);
-  if(lockstate == EV_ALREADY_OWNED) return;
+  if(lockstate == EV_ALREADY_OWNED) {
+    mtevL(eventer_deb, "Incoming event: %p already owned by this thread\n", e);
+    /* The incoming triggered event is already owned by this thread.  
+       This means our floated event completed before the current
+       event handler even exited.  So it retriggered recursively
+       from inside the event handler.  
+       
+       Treat this special case the same as a cross thread trigger
+       and just queue this event to be picked up on the next loop
+    */
+    eventer_cross_thread_trigger(e, mask);
```

Comment:
This approach will lead to this log:
```
    if(master_fds[fd].e != NULL) {
      mtevL(eventer_deb, "Attempting to trigger already-registered event fd: %d cross thread.\n", fd);
    }
```
because we've already set `master_fds[fd].e = e;` before calling cross_thread_trigger, so when we re-enter this handler cross-thread, `master_fds[fd].e != NULL`, and we hit that debug log. Given the (commented-out) `mtevAssert` in that code block, this seems contrary to intent.

What about changing the `if(master_fds[fd].e == NULL)` case above like follows:
```c
if(master_fds[fd].e == NULL) {
  lockstate = acquire_master_fd(fd);
  if (lockstate == EV_ALREADY_OWNED) {
    eventer_cross_thread_trigger(e,mask);
    return;
  }
  master_fds[fd].e = e;
  e->mask = 0;
  needs_add = 1;
}
```
?